lib/src/Uncertainty/Algorithm/Optimization/EfficientGlobalOptimization.cxx

```diff
@@ -129,43 +152,47 @@ class ExpectedImprovementEvaluation : public EvaluationImplementation
 
   UnsignedInteger getInputDimension() const
   {
-    return metaModelResult_.getMetaModel().getInputDimension();
+    if (isGPR_)
+      return gprResult_.getMetaModel().getInputDimension();
+    else
+      return krigingResult_.getMetaModel().getInputDimension();
   }
 
   UnsignedInteger getOutputDimension() const
   {
     return 1;
   }
 
-  Description getInputDescription() const
-  {
-    return metaModelResult_.getMetaModel().getInputDescription();
-  }
-
-  Description getOutputDescription() const
-  {
-    return metaModelResult_.getMetaModel().getOutputDescription();
-  }
-
 protected:
   Scalar optimalValue_;
-  KrigingResult metaModelResult_;
+  KrigingResult krigingResult_;
+  GaussianProcessRegressionResult gprResult_;
+  GaussianProcessConditionalCovariance gprCov_;
+  Bool isGPR_ = false;
   Function noiseModel_;
 
   // whether the global problem is a miminization (the improvement criterion is always maximized)
   Bool isMinimization_ = true;
 };
 
 
-
-
 void EfficientGlobalOptimization::run()
 {
   const OptimizationProblem problem(getProblem());
   const UnsignedInteger dimension = problem.getDimension();
   const Function model(problem.getObjective());
-  Sample inputSample(krigingResult_.getInputSample());
-  Sample outputSample(krigingResult_.getOutputSample());
+  Sample inputSample;
+  Sample outputSample;
+  if (isGPR_)
+  {
+    inputSample = gprResult_.getInputSample();
+    outputSample = gprResult_.getOutputSample();
+  }
+  else
+  {
+    inputSample = krigingResult_.getInputSample();
+    outputSample = krigingResult_.getOutputSample();
+  }
   UnsignedInteger size = inputSample.getSize();
   Point noise(size);
   const Bool hasNoise = metamodelNoise_.getEvaluation().getImplementation()->isActualImplementation();
```

Comment:
In the GPR case, we could detect a positive nugget factor, which can represent a homoscedastic noise (equivalent to a constant noise function).

Something like this:
```suggestion
  const Bool hasNoise = gprResult_.getCovarianceModel().getNuggetFactor() > ResourceMap::GetAsScalar("CovarianceModel-DefaultNuggetFactor");
```

Note that the [only EGO example in the doc with a noisy function](https://openturns.github.io/openturns/master/auto_numerical_methods/optimization/plot_ego.html?highlight=noise+function) uses a constant noise function, so this example could easily be converted to the new API.